metadata/src/test/java/org/apache/kafka/image/ScramImageTest.java

```diff
@@ -0,0 +1,161 @@
+        ScramCredentialData secondAlpha256Credential = randomScramCredentialData(random);
+        DELTA1_RECORDS.add(new ApiMessageAndVersion(new UserScramCredentialRecord().
+                setName("alpha").
+                setMechanism(SCRAM_SHA_256.type()).
+                setIterations(secondAlpha256Credential.iterations()).
+                setSalt(secondAlpha256Credential.salt()).
+                setSaltedPassword(secondAlpha256Credential.saltedPassword()), (short) 0));
+        DELTA1 = new ScramDelta(IMAGE1);
+        RecordTestUtils.replayAll(DELTA1, DELTA1_RECORDS);
+
+        Map<ScramMechanism, Map<String, ScramCredentialData>> image2mechanisms = new HashMap<>();
+
+        Map<String, ScramCredentialData> image2sha256 = new HashMap<>();
+        image2sha256.put("alpha", secondAlpha256Credential);
+        image2sha256.put("beta", image1sha256.get("beta"));
+        image2mechanisms.put(SCRAM_SHA_256, image2sha256);
+
+        Map<String, ScramCredentialData> image2sha512 = new HashMap<>();
+        image2sha512.put("alpha", image1sha512.get("alpha"));
+        image2mechanisms.put(SCRAM_SHA_512, image2sha512);
+
+        IMAGE2 = new ScramImage(image2mechanisms);
+    }
+
+    @Test
+    public void testEmptyImageRoundTrip() throws Throwable {
+        testToImageAndBack(ScramImage.EMPTY);
+    }
+
+    @Test
+    public void testImage1RoundTrip() throws Throwable {
+        testToImageAndBack(IMAGE1);
+    }
+
+    @Test
+    public void testApplyDelta1() throws Throwable {
+        assertEquals(IMAGE2, DELTA1.apply());
+    }
+
+    @Test
+    public void testImage2RoundTrip() throws Throwable {
+        testToImageAndBack(IMAGE2);
+    }
+
+    private void testToImageAndBack(ScramImage image) throws Throwable {
+        RecordListWriter writer = new RecordListWriter();
+        image.write(writer, new ImageWriterOptions.Builder().build());
+        ScramDelta delta = new ScramDelta(ScramImage.EMPTY);
+        RecordTestUtils.replayAll(delta, writer.records());
+        ScramImage nextImage = delta.apply();
+        assertEquals(image, nextImage);
+    }
+
+    @Test
+    public void testEmptyWithInvalidIBP() throws Throwable {
+        ImageWriterOptions imageWriterOptions = new ImageWriterOptions.Builder().
+                setMetadataVersion(MetadataVersion.IBP_3_4_IV0).build();
+        RecordListWriter writer = new RecordListWriter();
+        ScramImage.EMPTY.write(writer, imageWriterOptions);
+    }
+
+    @Test
+    public void testImage1withInvalidIBP() throws Throwable {
+        ImageWriterOptions imageWriterOptions = new ImageWriterOptions.Builder().
+                setMetadataVersion(MetadataVersion.IBP_3_4_IV0).build();
+        RecordListWriter writer = new RecordListWriter();
+        try {
+            IMAGE1.write(writer, imageWriterOptions);
+            fail("expected exception writing IMAGE with SCRAM records for MetadataVersion.IBP_3_4_IV0");
```

Comment:
You should be able to catch this with assertThrows and inspect the exception a bit. But we can do this in a follow-on.